protocol/x/subaccounts/keeper/isolated_subaccount.go

```diff
@@ -140,3 +138,193 @@ func isValidIsolatedPerpetualUpdates(
+	// If the size of the update and the position are the same, the perpetual update must have opened
+	// the position.
+	if perpetualUpdate.GetBigQuantums().Cmp(perpetualPosition.GetBigQuantums()) == 0 {
+		if len(settledUpdateWithUpdatedSubaccount.AssetUpdates) != 1 {
+			return nil, errorsmod.Wrapf(
+				types.ErrFailedToUpdateSubaccounts,
+				"Subaccount with id %v opened perpteual position with perpetual id %d with invalid number of"+
+					" changes to asset positions (%d), should only be 1 asset update",
+				updatedSubaccount.Id,
+				perpetualUpdate.PerpetualId,
+				len(settledUpdateWithUpdatedSubaccount.AssetUpdates),
+			)
+		}
+		if settledUpdateWithUpdatedSubaccount.AssetUpdates[0].AssetId != assettypes.AssetUsdc.Id {
+			return nil, errorsmod.Wrapf(
+				types.ErrFailedToUpdateSubaccounts,
+				"Subaccount with id %v opened perpteual position with perpetual id %d without a change to the"+
+					" quote currency's asset position.",
+				updatedSubaccount.Id,
+				perpetualUpdate.PerpetualId,
+			)
+		}
+		// Collateral equal to the quote currency asset position before the update was applied needs to be transferred.
+		// Subtract the delta from the updated subaccount's quote currency asset position size to get the size
+		// of the quote currency asset position.
+		quoteQuantumsBeforeUpdate := new(big.Int).Sub(
+			updatedSubaccount.GetUsdcPosition(),
+			settledUpdateWithUpdatedSubaccount.AssetUpdates[0].GetBigQuantums(),
+		)
+		return &types.IsolatedPerpetualPositionStateTransition{
+			SubaccountId:  updatedSubaccount.Id,
+			PerpetualId:   perpetualUpdate.PerpetualId,
+			QuoteQuantums: quoteQuantumsBeforeUpdate,
+			Transition:    types.Opened,
+		}, nil
+	}
+
+	// The isolated perpetual position changed size but was not opened or closed.
+	return nil, nil
+}
+
+// transferCollateralForIsolatedPerpetual transfers collateral between an isolated collateral pool
+// and the cross-perpetual collateral pool based on whether an isolated perpetual position was
+// opened or closed in a subaccount.
+// Note: This uses the `x/bank` keeper and modifies `x/bank` state.
+func (k *Keeper) transferCollateralForIsolatedPerpetual(
+	ctx sdk.Context,
+	stateTransition *types.IsolatedPerpetualPositionStateTransition,
+) error {
+	// No collateral to transfer if no state transition.
+	if stateTransition == nil {
+		return nil
+	}
+
+	isolatedCollateralPoolAddr, err := k.GetCollateralPoolFromPerpetualId(ctx, stateTransition.PerpetualId)
+	if err != nil {
+		return err
+	}
+	var toModuleAddr sdk.AccAddress
+	var fromModuleAddr sdk.AccAddress
+
+	// If an isolated perpetual position was opened in the subaccount, then move collateral from the
+	// cross-perpetual collateral pool to the isolated perpetual collateral pool.
+	if stateTransition.Transition == types.Opened {
+		toModuleAddr = isolatedCollateralPoolAddr
+		fromModuleAddr = types.ModuleAddress
+		// If the isolated perpetual position was closed, then move collateral from the isolated
```

Comment:
nit: fix indent

Reply:
That's from `go fmt`, this indent is expected. I tried to fix it too and `go fmt` kept on changing it back.

Reply:
weird